samples/AspNetCore/SecretStoreConfigurationProviderSample/Program.cs

```diff
@@ -0,0 +1,107 @@
+﻿namespace SecretStoreConfigurationProviderSample
+{
+    using Microsoft.AspNetCore.Hosting;
+    using Microsoft.Extensions.Hosting;
+    using Dapr.Client;
+    using Dapr.Extensions.Configuration;
+    using Microsoft.Extensions.DependencyInjection;
+    using System.Net.Sockets;
+    using System;
+    using System.Net;
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Secret Store Configuration Provider Sample.
+    /// </summary>
+    public class Program
+    {
+        const string localhost = "127.0.0.1";
+
+        static string daprPort => Environment.GetEnvironmentVariable("DAPR_GRPC_PORT") ?? "50001";
+
+        /// <summary>
+        /// Main for Secret Store Configuration Provider Sample.
+        /// </summary>
+        /// <param name="args">Arguments.</param>
+        public static void Main(string[] args)
+        {
+            CreateHostBuilder(args).Build().Run();
+        }
+
+        /// <summary>
+        /// Creates WebHost Builder.
+        /// </summary>
+        /// <param name="args">Arguments.</param>
+        /// <returns>Returns IHostbuilder.</returns>
+        public static IHostBuilder CreateHostBuilder(string[] args)
+        {
+            WaitForDapr();
+
+            // Create Dapr Client
+            var client = new DaprClientBuilder()
+                .Build();
+
+            return Host.CreateDefaultBuilder(args)
+                .ConfigureServices((services) =>
+                {
+                    // Add the DaprClient to DI.
+                    services.AddSingleton<DaprClient>(client);
+                })
+                .ConfigureAppConfiguration((configBuilder) =>
+                {
+                    // Create descriptors for the secrets you want to rerieve from the Dapr Secret Store.
+                    var secretDescriptors = new DaprSecretDescriptor[]
+                    {
+                        new DaprSecretDescriptor(
+                            "super-secret",
+                            new Dictionary<string, string>(){ { "namespace", "default" } }
+                        )
+                    };
+
+                    // Add the secret store Configuration Provider to the configuration builder.
+                    configBuilder.AddDaprSecretStore("kubernetes", secretDescriptors, client);
+                })
+                .ConfigureWebHostDefaults(webBuilder =>
+                {
+                    webBuilder.UseStartup<Startup>();
+                });
+        }
+
+        private static void WaitForDapr()
+        {
+            // Open the application port to trick Dapr into thinking it's ready 
+            var ipadress = IPAddress.Parse(localhost);
+            var server = new TcpListener(ipadress, 80);
+            server.Start();
```

Comment:
Looks good overall, few more questions: Is this server start still needed because you added a endpoint "/secret" to it, it can just use the WebHostBuilder?

Reply:
Even with the endpoint we cannot relay on WebHostBuilder cause using the configuration provider means that the app will attempt to load the secrets (through Dapr) before it's ready to listen (app port is not open yet in this stage). 

If we do not open the port, Dapr keeps waiting for the app which also keeps waiting for Dapr and we hit a dead lock.

Reply:
I am not too convinced on this workaround, Can it lazy load the secrets when needed and retry on exception. Starting this another server doesnt seem like the right thing to do.

Reply:
I see, if I add a retry loop for the secrets, the app will keep waiting forever. 

On the other hand I'm not sure how or if it's possible to implement lazy loading of configuration. Any ideas here?

Reply:
@rynowak Do you have any suggestion here?

Reply:
Hi @amanbha & @rynowak I came up with a reloading timer in the provider that could be a cleaner approach to this issue:

```csharp
public DaprSecretStoreConfigurationProvider(string store, IEnumerable<DaprSecretDescriptor> secretDescriptors, DaprClient client)
        {
            ArgumentVerifier.ThrowIfNullOrEmpty(store, nameof(store));
            ArgumentVerifier.ThrowIfNull(secretDescriptors, nameof(secretDescriptors));
            ArgumentVerifier.ThrowIfNull(client, nameof(client));

            this.store = store;
            this.secretDescriptors = secretDescriptors;
            this.client = client;

            reloadTimer = new System.Timers.Timer(1000);
            reloadTimer.Elapsed += (s, e) =>
            {
                reloadCount++;
                Console.WriteLine(reloadCount);
                try
                {
                    reloadTimer.Stop();
                    Load(true);
                }
                catch
                {
                    if (reloadCount > 5)
                    {
                        throw;
                    }
                    reloadTimer.Start();
                }
            };
            reloadTimer.Start();
        }

        public override void Load() => Load(false);

        private void Load(bool daprReady)
        {
            if (daprReady)
            {
                LoadAsync().ConfigureAwait(false).GetAwaiter().GetResult();
                OnReload();
            }
        }

        private async Task LoadAsync()
        {
            var data = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
            foreach (var secretDescriptor in secretDescriptors)
            {
                var result = await client.GetSecretAsync(store, secretDescriptor.SecretName, secretDescriptor.Metadata).ConfigureAwait(false);

                foreach (var key in result.Keys)
                {
                    if (data.ContainsKey(key))
                    {
                        throw new InvalidOperationException($"A duplicate key '{key}' was found in the secret store '{store}'. Please remove any duplicates from your secret store.");
                    }

                    data.Add(key, result[key]);
                }
            }

            Data = data;
        }
```

There are at least two problems with this approach:

1. Any error on the load method gets swallowed by the timer.
2. Secrets read through the provider may be unavailable for Startup.cs methods.

Thoughts?

Reply:
@cmendible Thsi looks like an acceptable workaround to me, lets see if @rynowak has a better solution from aspnetcore perspective.

Reply:
Hi @amanbha any news on this one?

Reply:
@cmendible I would say, please add this workaround.

Reply:
ok let me see I f I can come up with a solution so that the exceptions are not swallowed by the timer...

Reply:
Ok so if in the timer we do:

```csharp
...
catch
{
	if (reloadCount > 5)
	{
		System.Threading.ThreadPool.QueueUserWorkItem(_ => { throw ex; });
	}
	reloadTimer.Start();
}
...
```
the exception bubbles up to the main thread as needed. 

But the thing that keeps bothering me about this solution is that any secret needed in the `ConfigureServices` or `Configure` methods of the `Startup` class will not be available (i.e. a database connection string or password needed to register EF).

With the code as-is, based on the `WaitForDapr` method, users will not face the issue mentioned above.

Reply:
@cmendible I see, but I don't think that starting a new server is the right thing to do. Perhaps the way secrets are implemented, they are not the right candidate here to integrate with configuration provider and apps can query the secrets on demand when needed.

Reply:
@amanbha do you think we can explore the possibility of the secret api becoming available to the app container before dapr starts listening external calls? In ASP.NET Core we usually need connection strings and other secrets In the ConfigureServices of the Startup class. cc @yaron2

Reply:
> @amanbha do you think we can explore the possibility of the secret api becoming available to the app container before dapr starts listening external calls? In ASP.NET Core we usually need connection strings and other secrets In the ConfigureServices of the Startup class. cc @yaron2

Yes, that could be a possibility. @yaron2 whats your opninion on making secret api available in dapr sidecar before user app is ready?

Reply:
@cmendible I have opened a proposal for this in dapr repo https://github.com/dapr/dapr/issues/1493